### PR TITLE
Build R-CNN with ResNet

### PR DESCRIPTION
**Pull request by JihongJu**
- Build an RCNN model with keras_resnet.ResNet50.
 - Implement heads described in Mask R-CNN.

My current implementation is:
```python
        y = keras_resnet.ResNet50(inputs)
        features = y.layers[-2].output

        rpn_classification = keras.layers.Conv2D(
            9 * 1, (1, 1), activation="sigmoid")(features)
        rpn_regression = keras.layers.Conv2D(9 * 4, (1, 1))(features)

        rpn_prediction = keras.layers.concatenate(
            [rpn_classification, rpn_regression])

        # proposals of shape (1, None, 4)
        proposals = keras_rcnn.layers.object_detection.ObjectProposal(
            rois)([rpn_regression, rpn_classification])

        # slices of shape (1, None, 7, 7, 3)
        slices = keras_rcnn.layers.ROI((7, 7), rois)([inputs, proposals])
        
        # Implement the R-CNN heads shown in the figure below
        [score, boxes] = keras_rcnn.heads.ResHead(classes)(slices)
```
And the R-CNN heads from Mask R-CNN looks like
resnet_heads

That becomes:
```python
        y = keras.layers.TimeDistributed(
            keras.layers.Conv2D(1024, (1, 1)))(x)

        # conv5 block as in Deep Residual Networks with first conv operates
        # on a 7x7 RoI with stride 1 (instead of 14x14 / stride 2)
        for i in range(3):
            y = _bottleneck(512, (1, 1))(y)

        y = keras.layers.TimeDistributed(
            keras.layers.BatchNormalization(axis=channel_axis))(y)
        y = keras.layers.TimeDistributed(
            keras.layers.Activation("relu"))(y)

        # class and box branches
        y = keras.layers.TimeDistributed(
            keras.layers.AveragePooling2D((7, 7)))(y)

        score = keras.layers.TimeDistributed(
            keras.layers.Dense(classes, activation="softmax"))(y)

        boxes = keras.layers.TimeDistributed(
            keras.layers.Dense(4 * classes))(y)
```
in `keras_rcnn.heads.ResHead`.

What do you think? @0x00b1 

P.S. The API design should be discussed in broadinstitute/keras-rcnn#28

### CODE REVIEW

**Comment by 0x00b1** [comment]
@JihongJu Whoa! You’ve been busy! I’m excited to comment. Would you mind giving me a day or two? It’ll take some time to consider and respond.

**Comment by 0x00b1** [comment]
cc: @mcquin

File: keras_rcnn/heads/resnet.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+def ResHead(classes, mask=False):
+    """Resnet heads as in Mask R-CNN."""
+    def f(x):
+        if keras.backend.image_data_format() == "channels_last":
+            channel_axis = 3
+        else:
+            channel_axis = 1
+
+        y = keras.layers.TimeDistributed(
```
Comment (0x00b1) [inline keras_rcnn/heads/resnet.py]: In my opinion, this is the trickiest part of the implementation.

I foresee two problems with using TimeDistributed.

- I suspect that in nearly every situation TimeDistributed is used to step across a temporal dimension. Unfortunately, we’re the rare exception. It’ll be a challenge to explain this in our documentation (especially if we expect users to re-implement the CNN that use for feature extraction to use TimeDistributed). I’d really like Keras to modify the name to better emphasize its utility.

- TimeDistributed requires users to implement their feature extractor layer by layer instead of calling on a model instance directly. I’d really prefer something like:

```Python
x = keras.layers.Input((224, 224, 3))

y = keras_resnet.ResNet50(x)

y = keras.layers.TimeDistributed(y)
```

If it worked like this, we could hide the `y = keras.layers.TimeDistributed(y)` statement inside the RCNN constructor. We could probably implement our own wrapper that iterates across a model’s layers and appends a TimeDistributed layer.

File: keras_rcnn/layers/pooling.py [hunk 2]
```diff
@@ -32,15 +28,15 @@ def build(self, input_shape):
...
         # convert regions from (x, y, w, h) to (x1, y1, x2, y2)
-        regions = keras.backend.cast(regions, keras.backend.floatx())
-        regions = regions / self.stride
-        x1 = regions[:, 0]
-        y1 = regions[:, 1]
-        x2 = regions[:, 0] + regions[:, 2]
-        y2 = regions[:, 1] + regions[:, 3]
+        boxes = keras.backend.cast(boxes, keras.backend.floatx())
+        boxes = boxes / self.stride
+        x1 = boxes[..., 0]
```
Comment (0x00b1) [inline keras_rcnn/layers/pooling.py]: 👍

File: keras_rcnn/models.py [hunk 3]
```diff
@@ -1,33 +1,40 @@
 # -*- coding: utf-8 -*-
 
 import keras
+import keras_resnet
 import keras_rcnn.layers
+import keras_rcnn.heads
 
 
 class RCNN(keras.models.Model):
```
Comment (0x00b1) [inline keras_rcnn/models.py]: This looks great!

We should also consider adding a `features` argument that users would use to pass their feature extractor (e.g. ResNet50). What do you think?

Comment (JihongJu) [inline keras_rcnn/models.py]: @0x00b1 Actually, I think we should go for two arguments: features and heads (maybe rename these two to keep consistency).

File: keras_rcnn/heads/resnet.py (hunk 1 above)
Comment (JihongJu) [inline keras_rcnn/heads/resnet.py]: @0x00b1 I agree with you here. But I can hardly see how to implement such a wrapper. 
For example, we have `ResHead` that takes feature `x` as input and returns some outputs `y`. We need to apply TimeDistributed to each of the layers in between.
Maybe you have good ideas how to do that?

Comment (0x00b1) [inline keras_rcnn/heads/resnet.py]: Good question. 😆 

A quick method might be iterating over model.layers then pop each layer and append TimeDistributed.

**Comment by 0x00b1** [comment]
@JihongJu I merged the backend and pooling changes.

**Comment by JihongJu** [comment]
Great. I will continue the work on weekends.

File: keras_rcnn/heads/resnet.py (hunk 1 above)
Comment (0x00b1) [inline keras_rcnn/heads/resnet.py]: Hi, @JihongJu, I added a `temporal` module to keras-resnet. You should be able to use `keras_resnet.block.temporal.basic`, `keras_resnet.block.temporal.bottleneck`, and `keras_resnet.block.temporal._shortcut`.

Comment (JihongJu) [inline keras_rcnn/heads/resnet.py]: @0x00b1 That would be great! I am a bit busy these days. I will get back to this asap.

Comment (JihongJu) [inline keras_rcnn/heads/resnet.py]: @0x00b1  I sent a PR to keras-resnet to fix a bug for the temporal shortcut. And could you make a release (0.0.3?) afterwards?

Comment (0x00b1) [inline keras_rcnn/heads/resnet.py]: Yep :)

Comment (0x00b1) [inline keras_rcnn/heads/resnet.py]: Released (0.0.4 because I made a dumb release mistake!)

https://pypi.python.org/pypi/keras-resnet/0.0.4

File: keras_rcnn/models.py (hunk 3 above)
Comment (0x00b1) [inline keras_rcnn/models.py]: @JihongJu I like it! 👍

File: keras_rcnn/heads/resnet.py (hunk 1 above)
Comment (JihongJu) [inline keras_rcnn/heads/resnet.py]: @0x00b1 Great! Thanks!

**Comment by JihongJu** [comment]
@0x00b1 If you are okay with this design, I will continue with the Mask branch as in TODO with a new PR.

**Comment by 0x00b1** [comment]
@JihongJu Awesome! Merged! (I might do a little cleanup today or tomorrow.)
